File: 141-linked-list-cycle/141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution 
{
    public boolean hasCycle(ListNode head) 
    {
        if(head == null)
            return false;
        
        ListNode slow = head;
        ListNode fast = head;
        
        while(fast.next != null)
        {
            fast = fast.next;
            
            if(fast.next == null)
                return false;
            fast = fast.next;
            
            if(fast == slow)
                return true;
            
            slow = slow.next;
        }
        
        return false;
    }
}

File: 141-linked-list-cycle/141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    bool hasCycle(ListNode *head) 
    {
        if(head == nullptr)
            return false;
            
        ListNode* fast = head;
        
        while(fast->next != nullptr)
        {
            fast = fast->next;
            if(fast->next == nullptr)
                return false;
            fast = fast->next;
            head = head->next;
            
            if(head == fast)
                return true;
        }
        
        return false;
    }
};